File: rlm/cli/tui/channel_console.py
"""Channel Console — painel Rich de awareness multichannel para o TUI.

Exibe status de todos os canais registrados no ChannelStatusRegistry,
seja via HTTP (live mode) ou acesso direto ao CSR (local mode).
Permite envio cross-channel e probe sob demanda sem depender do REPL.
"""

from __future__ import annotations

import time
from dataclasses import dataclass, field
from typing import Any

from rich.console import Group, RenderableType
from rich.panel import Panel
from rich.table import Table
from rich.text import Text


# ── Status icons ──────────────────────────────────────────────────────────

_ICON_RUNNING = "●"
_ICON_CONFIGURED = "◑"
_ICON_OFF = "○"
_ICON_ERROR = "✖"


def _channel_icon(snap: dict[str, Any]) -> tuple[str, str]:
    """Retorna (icon, style) para um snapshot de canal."""
    if snap.get("running") and snap.get("healthy"):
        return _ICON_RUNNING, "bold green"
    if snap.get("running") and not snap.get("healthy"):
        return _ICON_ERROR, "bold red"
    if snap.get("configured"):
        return _ICON_CONFIGURED, "yellow"
    return _ICON_OFF, "dim"


# ── Data ──────────────────────────────────────────────────────────────────

@dataclass
class ChannelSnapshot:
    """Representação simplificada de um canal para renderização."""
    channel_id: str
    account_id: str = "default"
    configured: bool = False
    running: bool = False
    healthy: bool = False
    identity_name: str = ""
    last_error: str | None = None
    reconnect_attempts: int = 0
    last_probe_ms: float = 0.0
    meta: dict[str, Any] = field(default_factory=dict)

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> ChannelSnapshot:
        ident = d.get("identity") or {}
        name = ident.get("display_name") or ident.get("username") or ""
        return cls(
            channel_id=d.get("channel_id", "?"),
            account_id=d.get("account_id", "default"),
            configured=bool(d.get("configured")),
            running=bool(d.get("running")),
            healthy=bool(d.get("healthy")),
            identity_name=str(name),
            last_error=d.get("last_error"),
            reconnect_attempts=int(d.get("reconnect_attempts", 0)),
            last_probe_ms=float(d.get("last_probe_ms", 0)),
            meta=dict(d.get("meta") or {}),
        )


# ── Panel builder ─────────────────────────────────────────────────────────

@dataclass
class ChannelConsoleState:
    """Estado mantido entre renders do painel de canais."""
    snapshots: list[ChannelSnapshot] = field(default_factory=list)
    last_fetch_at: float = 0.0
    last_send_result: str = ""
    fetch_error: str = ""


def build_channel_panel(state: ChannelConsoleState) -> Panel:
    """Constrói o painel Rich de canais a partir do estado atual."""
    blocks: list[RenderableType] = []

    if not state.snapshots and not state.fetch_error:
        blocks.append(Text("Nenhum canal registrado.", style="dim"))
    elif state.fetch_error:
        blocks.append(Text(f"Erro: {state.fetch_error}", style="bold red"))
    else:
        table = Table(box=None, expand=True, show_header=True, padding=(0, 1))
        table.add_column("", width=2)  # icon
        table.add_column("Canal", width=11)
        table.add_column("Bot", ratio=1)
        table.add_column("Latência", width=8, justify="right")
        table.add_column("Erros", width=5, justify="right")

        for snap in state.snapshots:
            icon, style = _channel_icon(
                {"running": snap.running, "healthy": snap.healthy, "configured": snap.configured}
            )
            latency = f"{snap.last_probe_ms:.0f}ms" if snap.last_probe_ms > 0 else "-"
            errors = str(snap.reconnect_attempts) if snap.reconnect_attempts else "-"
            name = snap.identity_name or "-"
            if snap.last_error and not snap.healthy:
                name = Text(snap.last_error[:30], style="red")

            table.add_row(
                Text(icon, style=style),
                snap.channel_id,
                name if isinstance(name, Text) else str(name),
                latency,
                errors,
            )
        blocks.append(table)

    # Status line
    summary_parts: list[str] = []
    total = len(state.snapshots)
    running = sum(1 for s in state.snapshots if s.running)
    if total:
        summary_parts.append(f"{running}/{total} ativos")
    if state.last_fetch_at:
        age = time.time() - state.last_fetch_at
        summary_parts.append(f"atualizado há {age:.0f}s")
    if summary_parts:
        blocks.append(Text())
        blocks.append(Text(" · ".join(summary_parts), style="dim"))

    if state.last_send_result:
        blocks.append(Text(state.last_send_result, style="bold"))

    # Help
    blocks.append(Text())
    help_text = Text()
    help_text.append("/channels", style="bold cyan")
    help_text.append("  atualizar status\n")
    help_text.append("/send <canal> <texto>", style="bold cyan")
    help_text.append("  enviar cross-channel\n")
    help_text.append("/probe <canal>", style="bold cyan")
    help_text.append("  testar conectividade")
    blocks.append(help_text)

    return Panel(
        Group(*blocks),
        title="Canais",
        border_style="magenta",
    )


# ── Data fetchers ─────────────────────────────────────────────────────────

def fetch_channel_snapshots_live(live_api: Any) -> list[ChannelSnapshot]:
    """Busca snapshots via HTTP /api/channels/status (live mode)."""
    data = live_api.fetch_channels_status()
    channels = data.get("channels") or {}
    result: list[ChannelSnapshot] = []
    for _channel_id, accounts in channels.items():
        if isinstance(accounts, list):
            for acc in accounts:
                result.append(ChannelSnapshot.from_dict(acc))
        elif isinstance(accounts, dict):
            result.append(ChannelSnapshot.from_dict(accounts))
    return result


def fetch_channel_snapshots_local() -> list[ChannelSnapshot]:
    """Busca snapshots diretamente do CSR (local mode — se inicializado)."""
    try:
        from rlm.core.comms.channel_status import get_channel_status_registry
        csr = get_channel_status_registry()
    except (ImportError, RuntimeError):
        return []
    summary = csr.summary()
    channels = summary.get("channels") or {}
    result: list[ChannelSnapshot] = []
    for _channel_id, accounts in channels.items():
        if isinstance(accounts, list):
            for acc in accounts:
                result.append(ChannelSnapshot.from_dict(acc))
        elif isinstance(accounts, dict):
            result.append(ChannelSnapshot.from_dict(accounts))
    return result


def refresh_channel_state(
    state: ChannelConsoleState,
    *,
    live_api: Any | None,
) -> None:
    """Atualiza o state in-place — nunca levanta exceção."""
    try:
        if live_api is not None:
            state.snapshots = fetch_channel_snapshots_live(live_api)
        else:
            state.snapshots = fetch_channel_snapshots_local()
        state.last_fetch_at = time.time()
        state.fetch_error = ""
    except Exception as exc:
        state.fetch_error = str(exc)[:120]

File: packages/cli/src/tui/channel-console.ts
/**
 * Channel Console — painel de awareness multichannel para o TUI.
 *
 * Exibe status de todos os canais registrados, permite envio cross-channel
 * e probe sob demanda sem depender de bibliotecas de UI externas.
 *
 * Migrado de rlm/cli/tui/channel_console.py
 */

export interface ChannelStatusApi {
  fetchChannelsStatus(): Promise<Record<string, unknown>>;
}

// ─────────────────────────────────────────────── Ícones de status

const ICON_RUNNING = "●";
const ICON_CONFIGURED = "◑";
const ICON_OFF = "○";
const ICON_ERROR = "✖";

type IconEntry = { icon: string; style: "running" | "error" | "configured" | "off" };

function channelIcon(snap: {
  running: boolean;
  healthy: boolean;
  configured: boolean;
}): IconEntry {
  if (snap.running && snap.healthy) return { icon: ICON_RUNNING, style: "running" };
  if (snap.running && !snap.healthy) return { icon: ICON_ERROR, style: "error" };
  if (snap.configured) return { icon: ICON_CONFIGURED, style: "configured" };
  return { icon: ICON_OFF, style: "off" };
}

// ─────────────────────────────────────────────── Snapshot de canal

export interface ChannelSnapshot {
  channelId: string;
  accountId: string;
  configured: boolean;
  running: boolean;
  healthy: boolean;
  identityName: string;
  lastError: string | null;
  reconnectAttempts: number;
  lastProbeMs: number;
  meta: Record<string, unknown>;
}

export function channelSnapshotFromDict(d: Record<string, unknown>): ChannelSnapshot {
  const ident = (d["identity"] ?? {}) as Record<string, unknown>;
  const name = String(ident["display_name"] ?? ident["username"] ?? "");
  return {
    channelId: String(d["channel_id"] ?? "?"),
    accountId: String(d["account_id"] ?? "default"),
    configured: Boolean(d["configured"]),
    running: Boolean(d["running"]),
    healthy: Boolean(d["healthy"]),
    identityName: name,
    lastError: d["last_error"] != null ? String(d["last_error"]) : null,
    reconnectAttempts: Number(d["reconnect_attempts"] ?? 0),
    lastProbeMs: Number(d["last_probe_ms"] ?? 0),
    meta: (d["meta"] ?? {}) as Record<string, unknown>,
  };
}

// ─────────────────────────────────────────────── Estado do painel

export interface ChannelConsoleState {
  snapshots: ChannelSnapshot[];
  lastFetchAt: number;
  lastSendResult: string;
  fetchError: string;
}

export function createChannelConsoleState(): ChannelConsoleState {
  return {
    snapshots: [],
    lastFetchAt: 0,
    lastSendResult: "",
    fetchError: "",
  };
}

// ─────────────────────────────────────────────── Renderização em texto

/** Estilo ANSI mínimo para exibição no terminal sem dependências externas. */
const ANSI = {
  reset: "\x1b[0m",
  bold: "\x1b[1m",
  dim: "\x1b[2m",
  green: "\x1b[32m",
  yellow: "\x1b[33m",
  red: "\x1b[31m",
  magenta: "\x1b[35m",
  cyan: "\x1b[36m",
};

function styleIcon(entry: IconEntry): string {
  switch (entry.style) {
    case "running": return `${ANSI.bold}${ANSI.green}${entry.icon}${ANSI.reset}`;
    case "error": return `${ANSI.bold}${ANSI.red}${entry.icon}${ANSI.reset}`;
    case "configured": return `${ANSI.yellow}${entry.icon}${ANSI.reset}`;
    default: return `${ANSI.dim}${entry.icon}${ANSI.reset}`;
  }
}

/**
 * Constrói uma representação em texto do painel de canais.
 * Retorna um array de linhas prontas para `process.stdout.write`.
 */
export function buildChannelPanelLines(state: ChannelConsoleState): string[] {
  const lines: string[] = [];
  const border = `${ANSI.magenta}─────────────────────────────── Canais ───────────────────────────────${ANSI.reset}`;

  lines.push(border);

  if (!state.snapshots.length && !state.fetchError) {
    lines.push(`${ANSI.dim}  Nenhum canal registrado.${ANSI.reset}`);
  } else if (state.fetchError) {
    lines.push(`${ANSI.bold}${ANSI.red}  Erro: ${state.fetchError}${ANSI.reset}`);
  } else {
    // Cabeçalho
    lines.push(
      `  ${ANSI.bold}${"".padEnd(2)} ${"Canal".padEnd(12)} ${"Bot".padEnd(24)} ${"Lat".padStart(7)} ${"Err".padStart(4)}${ANSI.reset}`
    );

    for (const snap of state.snapshots) {
      const entry = channelIcon({
        running: snap.running,
        healthy: snap.healthy,
        configured: snap.configured,
      });
      const icon = styleIcon(entry);
      const latency = snap.lastProbeMs > 0 ? `${snap.lastProbeMs.toFixed(0)}ms` : "-";
      const errors = snap.reconnectAttempts ? String(snap.reconnectAttempts) : "-";
      let name = snap.identityName || "-";
      if (snap.lastError && !snap.healthy) {
        name = `${ANSI.red}${snap.lastError.slice(0, 24)}${ANSI.reset}`;
      }
      lines.push(
        `  ${icon} ${snap.channelId.padEnd(12)} ${name.padEnd(24)} ${latency.padStart(7)} ${errors.padStart(4)}`
      );
    }
  }

  // Linha de status
  const total = state.snapshots.length;
  const running = state.snapshots.filter((s) => s.running).length;
  const summaryParts: string[] = [];
  if (total) summaryParts.push(`${running}/${total} ativos`);
  if (state.lastFetchAt) {
    const age = Math.round((Date.now() / 1000) - state.lastFetchAt);
    summaryParts.push(`atualizado há ${age}s`);
  }
  if (summaryParts.length) {
    lines.push("");
    lines.push(`  ${ANSI.dim}${summaryParts.join(" · ")}${ANSI.reset}`);
  }

  if (state.lastSendResult) {
    lines.push(`  ${ANSI.bold}${state.lastSendResult}${ANSI.reset}`);
  }

  // Ajuda
  lines.push("");
  lines.push(
    `  ${ANSI.cyan}/channels${ANSI.reset}  atualizar status   ` +
    `${ANSI.cyan}/send <canal> <texto>${ANSI.reset}  cross-channel   ` +
    `${ANSI.cyan}/probe <canal>${ANSI.reset}  testar`
  );
  lines.push(border);

  return lines;
}

/**
 * Imprime o painel de canais no stdout.
 */
export function printChannelPanel(state: ChannelConsoleState): void {
  process.stdout.write(buildChannelPanelLines(state).join("\n") + "\n");
}

// ─────────────────────────────────────────────── Busca de snapshots

function extractSnapshots(data: Record<string, unknown>): ChannelSnapshot[] {
  const channels = (data["channels"] ?? {}) as Record<string, unknown>;
  const result: ChannelSnapshot[] = [];
  for (const accounts of Object.values(channels)) {
    if (Array.isArray(accounts)) {
      for (const acc of accounts) {
        result.push(channelSnapshotFromDict(acc as Record<string, unknown>));
      }
    } else if (typeof accounts === "object" && accounts !== null) {
      result.push(channelSnapshotFromDict(accounts as Record<string, unknown>));
    }
  }
  return result;
}

/** Busca snapshots via HTTP /api/channels/status (live mode). */
export async function fetchChannelSnapshotsLive(
  liveApi: ChannelStatusApi
): Promise<ChannelSnapshot[]> {
  const data = await liveApi.fetchChannelsStatus();
  return extractSnapshots(data);
}

/**
 * Atualiza o estado in-place — nunca lança exceção.
 * Suporta modo live (via LiveWorkbenchAPI) ou modo local (sem API).
 */
export async function refreshChannelState(
  state: ChannelConsoleState,
  opts: { liveApi?: ChannelStatusApi }
): Promise<void> {
  try {
    if (opts.liveApi) {
      state.snapshots = await fetchChannelSnapshotsLive(opts.liveApi);
    } else {
      // local mode: sem acesso ao Python backend — retorna vazio
      state.snapshots = [];
    }
    state.lastFetchAt = Date.now() / 1000;
    state.fetchError = "";
  } catch (err) {
    state.fetchError = String(err).slice(0, 120);
  }
}
